tools/testbench/topology_ipc4.c

```diff
@@ -0,0 +1,1304 @@
+	if (ret < 0)
+		return ret;
+
+	tplg_ctl = calloc(ctx->hdr->payload_size, 1);
+	if (!tplg_ctl) {
+		free(comp_info->ipc_payload);
+		return -ENOMEM;
+	}
+
+	/* only base config supported for now. extn support will be added later */
+	comp_info->ipc_size = sizeof(struct ipc4_base_module_cfg);
+	comp_info->ipc_payload = calloc(comp_info->ipc_size, 1);
+	if (!comp_info->ipc_payload)
+		return -ENOMEM;
+
+	/* FIXME: move this to when the widget is actually set up */
+	comp_info->instance_id = tp->instance_ids[SND_SOC_TPLG_DAPM_EFFECT]++;
+	comp_info->module_id = 0x9e; /* dcblock */
+
+	/* skip kcontrols for now, set object to NULL */
+	ret = tplg_create_controls(ctx, ctx->widget->num_kcontrols, tplg_ctl,
+				   ctx->hdr->payload_size, NULL);
+	if (ret < 0) {
+		fprintf(stderr, "error: loading controls\n");
+		goto out;
+	}
+
+	tb_setup_widget_ipc_msg(comp_info);
+
+	/* TODO: drop tplg_ctl to avoid memory leak. Need to store and handle this
+	 * to support controls.
+	 */
+	free(tplg_ctl);
+	return 0;
+
+out:
+	free(tplg_ctl);
+	free(comp_info->ipc_payload);
+	return ret;
+}
+
+static int tb_register_graph(struct testbench_prm *tp, int count)
+{
+	struct tplg_context *ctx = &tp->tplg;
+	int ret = 0;
+	int i;
+
+	for (i = 0; i < count; i++) {
+		ret = tplg_parse_graph(ctx, &tp->widget_list, &tp->route_list);
+		if (ret < 0)
+			return ret;
+	}
+
+	return ret;
+}
+
+static int tb_parse_pcm(struct testbench_prm *tp, int count)
+{
+	struct tplg_context *ctx = &tp->tplg;
+	int ret, i;
+
+	for (i = 0; i < count; i++) {
+		ret = tplg_parse_pcm(ctx, &tp->widget_list, &tp->pcm_list);
+		if (ret < 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * create a list with all widget info
+ * containing mapping between component names and ids
+ * which will be used for setting up component connections
+ */
+static inline int tb_insert_comp(struct testbench_prm *tp)
+{
+	struct tplg_context *ctx = &tp->tplg;
+	struct tplg_comp_info *comp_info;
+	int comp_id = ctx->comp_id;
+	int ret;
+
+	if (ctx->widget->id == SND_SOC_TPLG_DAPM_SCHEDULER)
+		return 0;
+
+	comp_info = calloc(sizeof(struct tplg_comp_info), 1);
+	if (!comp_info)
+		return -ENOMEM;
+
+	comp_info->name = strdup(ctx->widget->name);
+	if (!comp_info->name) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	comp_info->stream_name = strdup(ctx->widget->sname);
+	if (!comp_info->stream_name) {
+		ret = -ENOMEM;
+		goto sname_err;
+	}
+
+	comp_info->id = comp_id;
+	comp_info->type = ctx->widget->id;
+	comp_info->pipeline_id = ctx->pipeline_id;
+	ctx->current_comp_info = comp_info;
+
+	ret = tb_parse_ipc4_comp_tokens(tp, &comp_info->basecfg);
+	if (ret < 0)
+		goto sname_err;
+
+	list_item_append(&comp_info->item, &tp->widget_list);
+
+	printf("debug: loading comp_id %d: widget %s type %d size %d at offset %ld is_pages %d\n",
+	       comp_id, ctx->widget->name, ctx->widget->id, ctx->widget->size,
+	       ctx->tplg_offset, comp_info->basecfg.is_pages);
+
+	return 0;
+
+sname_err:
+	free(comp_info->name);
+
+err:
```

Comment:
one label at line 897 is enough - `free(NULL)` is allowed

Reply:
yep